### PR TITLE
pkg/policy: remove ReverseRules from Consumable

### LINKED ISSUE #2795: remove 'ReverseRules' from Consumable
Example:
If we want to allow identity B to have ingress traffic from identity A, then without connection tracking enabled, we have to allow ingress traffic to A from B as well so that A can receive replies from B - traffic must be allowed in both directions. As such, the structures within Cilium for policy have to know that identity B is allowed to receive traffic from identity A, and vice versa.

If ConnTrack is disabled for the endpoints, within the Consumables that apply to those endpoints, we have to find a way to specify which identities have been specified as being allowed to talk to each other. The current implementation in the Consumable structure is to have a map of security identities (ReverseRules) which specify which identities have specified that they want to receive traffic from that Consumable; the Consumable must know that it must allow traffic in the reverse direction so the connection can be maintained. However, this has a variety of issues associated with it:

* We enable ConnTrack for an endpoint dependent upon the policy:

`pkg/endpoint/policy.go:regeneratePolicy`:
```
// Apply possible option changes before regenerating maps, as map regeneration
	// depends on the conntrack options
	if c.L4Policy != nil {
		if c.L4Policy.RequiresConntrack() {
			opts[OptionConntrack] = optionEnabled
		}
	}
```

We thus change something that the user may have configured based on the policy. In my opinion, we should not be changing options that are configurable unless the user specifies them. 

* This does not work in a multinode setup 

The only way that the ReverseRules map is updated is based off of the `ConsumableCache`, which maps identities to Consumables on the local host. However, if the cache does not contain an entry for the identity for which the reverse rule is being added, as is the case when both identities are on separate nodes in the system, then no such reverse rule is added. This will always be the case when ConnTrack is disabled for two endpoints that are on separate nodes, as the ConsumableCache is only updated with identities local to the host on which the cilium-agent is running.

* This hasn't been working anyway upon changes in policy after the reverse rules have been updated: 

in `pkg/policy/consumer.go:BanConsumerLocked`, which is called when an identity is no longer allowed ingress access for a Consumable, there is the following logic:
```
if consumer.Reverse != nil {
	c.deleteReverseRule(id, c.ID)
}
```

However, the `Reverse` field for a Consumer is never populated anywhere in the code. Thus, in the case where endpoints are running on both nodes with ConnTrack disabled, the deletion of an identity from the `ReverseRules` of a Consumable wasn't happening when it should. 

I think this warrants us making this option immutable for an endpoint, or at least specifying clearly in our documentation in which cases this will work / not work (single node setup vs multinode setup). 

Feel free correct me if my assumptions are incorrect below :-)

### PR DESCRIPTION
**Pull request by ianvernon**
If connection-tracking was not enabled for an endpoint, in order to ensure that
traffic between endpoints managed by Cilium would flow, if a policy allowed
ingress to endpoint A from endpoint B, then B's policy maps were also updated
to allow ingress to itself from A as well. However, this only worked if both
endpoints were local to the same node, as the sharing of this information is
done on a per-daemon level; security identities are not resolved on each node
in the cluster. As such, remove ReverseRules, which was used to store such data,
from the Consumable structure.

There is a special case involving the reserved-identities, whose corresponding
Consumables do need to be updated if an endpoint allows traffic to them. 

Signed-off by: Ian Vernon <ian@cilium.io>

Fixes: #2795

### CODE REVIEW

**Comment by ianvernon** [comment]
test-me-please

**Comment by ianvernon** [comment]
test-me-please

**Comment by ianvernon** [comment]
So because this branch was rebased on top of another PR, it added a ton of CODEOWNERS to it, which unfortunately cannot be removed. The only ones that are needed for review on this PR are:
* policy
* agent
* endpoint

All reviews are welcome, though :)

**Review by joestringer — DISMISSED** [review]
LGTM, less cases to worry about.

CT can still be disabled with this patch, but users must disable CT for all endpoints that intend to still be able to talk to one another. This should only be used if you know what you're doing.

**Review by aanm — DISMISSED** [review]

**Comment by aanm** [comment]
This should brake some tests no?

**Comment by ianvernon** [comment]
@aanm which tests? the build passed when I ran it last time.

I'll go ahead and run `01-ct.sh` locally to see if this breaks it, as I suspect that that is the one you are referring to. This test was marked as validated / deprecated accordingly as part of #1841.

**Comment by aanm** [comment]
@ianvernon exactly. if it's marked as validated I'm hoping that ginkgo test should fail, no?

**Comment by jrajahalme** [comment]
> On Feb 14, 2018, at 3:21 PM, André Martins <notifications@github.com> wrote:
> 
> This should brake some tests no?
> 
> 

Not sure if I’m out-of-context, but even when conntrack has been disabled for an endpoint, policy calculation has enabled it if the L4 policy requires connection tracking (see policy.RequiresConntrack()). Put the other way, even tests with conntrack disabled have actually run conntrack enabled is the policy would not work otherwise.

  Jarno

**Comment by aanm** [comment]
@jrajahalme I wasn't aware of the `RequiresConntrack`. Thanks

**Comment by ianvernon** [comment]
@aanm, you were right; this lead me to find out an issue with the corresponding Ginkgo test, which I've fixed here: https://github.com/cilium/cilium/pull/2847 . 

#2847 should block this PR, as I will have update the tests accordingly on this PR (#2804) to account for the removal of ReverseRules.

**Comment by ianvernon** [comment]
test-me-please

File: pkg/policy/consumer.go
Comment (aanm) [inline pkg/policy/consumer.go:198]: Why is the logic exactly the same on both branches?

File: pkg/policy/consumer.go
Comment (aanm) [inline pkg/policy/consumer.go:233]: same here?

File: pkg/policy/identity.go
Comment (aanm) [inline pkg/policy/identity.go]: why not `_, e := ReservedIdentityNames[id]; return e`?

**Review by aanm — DISMISSED** [review]

File: pkg/policy/consumer.go
Comment (ianvernon) [reply pkg/policy/consumer.go:198]: It is not the same; one acquires and releases the lock, and one does not. If a Consumable can allow itself (which it can IIRC), then we should not call the lock for it because it will already be locked by the caller of `AllowIngressIdentityLocked`.

File: pkg/policy/identity.go
Comment (ianvernon) [reply pkg/policy/identity.go]: Heh, good idea. Will fix in next upload.

File: pkg/policy/consumer.go
Comment (ianvernon) [reply pkg/policy/consumer.go:233]: Same as above reply in `AllowIngressIdentityLocked`.

File: pkg/policy/identity.go
Comment (ianvernon) [reply pkg/policy/identity.go]: Fixed.

**Comment by ianvernon** [comment]
test-me-please

File: pkg/policy/consumer.go
Comment (aanm) [reply pkg/policy/consumer.go:198]: Please add this comment in the code

**Review by jrajahalme — APPROVED** [review]
Formal approval for proxy, did not read any code :-)

File: pkg/policy/consumer.go
Comment (ianvernon) [reply pkg/policy/consumer.go:198]: Done.

**Comment by ianvernon** [comment]
test-me-please

File: pkg/policy/consumer.go
Comment (ianvernon) [reply pkg/policy/consumer.go:198]: Sorry, now it is done - didn't push to origin for the branch before I commented.

**Comment by ianvernon** [comment]
test-me-please

File: test/runtime/connectivity.go
Comment (aanm) [inline test/runtime/connectivity.go:325]: Just to be sure, is this a leftover?

**Review by aanm — APPROVED** [review]

File: test/runtime/connectivity.go
Comment (ianvernon) [reply test/runtime/connectivity.go:325]: I left this because it clearly displays what policies are imported in Cilium to console without wrapping it in anything else - I can remove it if you like, though.

**Review by tgraf — APPROVED** [review]

**Pull request merged**
